src/lib/zkapp.ts

```diff
@@ -93,84 +110,277 @@ function method<T extends SmartContract>(
+                // proofs actually don't have to be cloned
+                previousProofs: getPreviousProofsForProver(
+                  actualArgs,
+                  methodIntf
+                ),
+                ZkappClass,
+                memoized,
+                blindingValue,
+              });
+            }
+            return result;
+          }
         }
-      );
-      return result;
-    } else if (!Mina.currentTransaction.has()) {
-      // outside a transaction, just call the method, but check precondition invariants
-      let result = method.apply(this, actualArgs);
-      // check the self party right after calling the method
-      // TODO: this needs to be done in a unified way for all parties that are created
-      assertPreconditionInvariants(this.self);
-      cleanPreconditionsCache(this.self);
-      assertStatePrecondition(this);
-      return result;
-    } else {
-      // in a transaction, also add a lazy proof to the self party
-      // (if there's no other authorization set)
-
-      // first, clone to protect against the method modifying arguments!
-      // TODO: double-check that this works on all possible inputs, e.g. CircuitValue, snarkyjs primitives
-      let clonedArgs = cloneCircuitValue(actualArgs);
-      let result = method.apply(this, actualArgs);
-      assertStatePrecondition(this);
-      let party = this.self;
-      if (!Authorization.hasAny(party)) {
-        Authorization.setLazyProof(party, {
-          method,
-          args: clonedArgs,
-          // proofs actually don't have to be cloned
-          previousProofs: getPreviousProofsForProver(actualArgs, methodIntf),
-          ZkappClass,
-        });
-      }
-      return result;
+      )[1];
     }
+    // if we're here, this method was called inside _another_ smart contract method
+    let parentParty = smartContractContext.get().this.self;
+    let methodCallDepth = smartContractContext.get().methodCallDepth;
+    let [, result] = smartContractContext.runWith(
+      { this: this, methodCallDepth: methodCallDepth + 1 },
+      () => {
+        // if the call result is not undefined but there's no known returnType, the returnType was probably not annotated properly,
+        // so we have to explain to the user how to do that
+        let { returnType } = methodIntf;
+        let noReturnTypeError =
+          `To return a result from ${methodIntf.methodName}() inside another zkApp, you need to declare the return type.\n` +
+          `This can be done by annotating the type at the end of the function signature. For example:\n\n` +
+          `@method ${methodIntf.methodName}(): Field {\n` +
+          `  // ...\n` +
+          `}\n\n` +
+          `Note: Only types built out of \`Field\` are valid return types. This includes snarkyjs primitive types and custom CircuitValues.`;
+        // if we're lucky, analyzeMethods was already run on the callee smart contract, and we can catch this error early
```

Comment:
Clarifying question. When does analyzeMethods run for the first time on a contract?

Reply:
It runs
* in `SmartContract.compile()` (before actually compiling)
* when computing the `SmartContract.digest()`
* in a `Mina.transaction` block, it runs if one of the methods needs it under the hood; this is only the case when reducing sequence events right now.
So, it definitely runs when we're proving things (because we `compile` first), but likely not if we don't.

The code above will throw the helpful error message if we're not proving, so that case is covered.

The case that's problematic is when we compile the caller first, and the callee second. Then, the "if we're lucky" case above is not true in the caller circuit. Then, we won't be able to nicely catch the missing return type annotation, and the caller's `compile` will throw an ugly, unhelpful error instead.

Thanks for poking into this :D Maybe I should make it so that analyzeMethods figures out all the SmartContracts called inside a method, and also runs analyzeMethods on them. This way, the callee would definitely be analyzed already in the code above, so we'd avoid the "ugly error" case.

Reply:
Update: I tried this out, but wasn't able to make it work. The ugly error is already thrown during analyzeMethods, so I have no way of finding the list of callees before. I think we'll have to live with a suboptimal error message in this edge case

Reply:
> Update: I tried this out, but wasn't able to make it work. The ugly error is already thrown during analyzeMethods, so I have no way of finding the list of callees before. I think we'll have to live with a suboptimal error message in this edge case

At least we are aware of the edge case error behavior.